cmd/rqlite/http/client.go

```diff
@@ -0,0 +1,186 @@
+
+// ErrTooManyRedirects indicates that the client exceeded the maximum number of redirects
+var ErrTooManyRedirects = fmt.Errorf("maximum leader redirect limit exceeded")
+
+// HostChangedError indicates that the underlying request was executed on a different host
+// different from the caller anticipated
+type HostChangedError struct {
+	NewHost string
+}
+
+func (he *HostChangedError) Error() string {
+	return fmt.Sprintf("HostChanedErr: new host is '%s'", he.NewHost)
+}
+
+type ConfigFunc func(*Client)
+
+// Client is a wrapper around stock http.Client that adds "retry on another host" behaviour
+// based on the supplied configuration.
+//
+// Errors that occur while performing requests are returned as they are, and the predicate to determine whether to check
+// another host or not is the request returning an `http.StatusServiceUnavailable`.
+// Note:
+//
+// This type is not goroutine safe.
+type Client struct {
+	*http.Client
+	scheme string
+	hosts  []string
+	Prefix string
+
+	// creds stores the http basic authentication username and password
+	creds  string
+	logger *log.Logger
+
+	// currentHost keeps track of the last available host
+	currentHost int
+	maxRedirect int
+}
+
+func NewClient(client *http.Client, hosts []string, configFuncs ...ConfigFunc) *Client {
+	cl := &Client{
+		Client:      client,
+		hosts:       hosts,
+		scheme:      "http",
+		maxRedirect: 21,
+		Prefix:      "/",
+		logger:      log.New(os.Stderr, "[client]", log.LstdFlags),
+	}
+
+	for _, f := range configFuncs {
+		f(cl)
+	}
+
+	return cl
+}
+
+// WithScheme changes the default scheme used i.e "http"
+func WithScheme(scheme string) ConfigFunc {
+	return func(client *Client) {
+		client.scheme = scheme
+	}
+}
+
+func WithPrefix(prefix string) ConfigFunc {
+	return func(client *Client) {
+		client.Prefix = prefix
+	}
+}
+
+// WithLogger changes the default logger to the one provided
+func WithLogger(logger *log.Logger) ConfigFunc {
+	return func(client *Client) {
+		client.logger = logger
+	}
+}
+
+// WithBasicAuth adds basic authentication behaviour to the client's request
+func WithBasicAuth(creds string) ConfigFunc {
+	return func(client *Client) {
+		client.creds = creds
+	}
+}
+
+func (c *Client) Query(url url.URL) (*http.Response, error) {
+	return c.execRequest(http.MethodGet, url, nil)
+}
+
+func (c *Client) Execute(url url.URL, body io.Reader) (*http.Response, error) {
+	return c.execRequest(http.MethodPost, url, body)
+}
+
+func (c *Client) execRequest(method string, url url.URL, body io.Reader) (*http.Response, error) {
+	triedHosts := 0
+	for triedHosts < len(c.hosts) {
+		host := c.hosts[c.currentHost]
+		url.Scheme = c.scheme
+		url.Host = host
+		urlStr := url.String()
+		resp, err := c.requestFollowRedirect(method, urlStr, body)
+
+		// If the status code is anything beside "service unavailable"
+		// we should propagate the error back to the caller
+		if resp != nil && resp.StatusCode != http.StatusServiceUnavailable {
```

Comment:
I'm not sure about this. Why not try another node if this node returns StatusServiceUnavailable? What's special about StatusServiceUnavailable?

Reply:
We want to try another server only if the node is unreachable, if the request itself is not well-formatted (and caused the server to return an error say `500`), trying on another server wouldn't be any different.

Reply:
I see your point. However I'm not sure if StatusServiceUnavailable is that special. I would need to check the code.

In fact, the issue associated with this PR states that the "try another node logic" should kick in if the first node "goes down". I interpret that to mean if the node is not reachable. But receiving any HTTP status means the node *is reachable* and we shouldn't try another node (except in the case of an explicit redirect). Does that make sense? Or do you think of it differently? If so, what your definition of "goes down"?

Reply:
So I retested my hypothesis, I initially assumed that if a node is down the client would wrap the "connection refused" into something like ServiceUnavailable, this is not actually the case, if a server is down (killed, oomed, evicted ...) a connection refused error is going to be returned with a response of nil.

So I guess the check for the specific ServiceUnavailable is not required.

Reply:
Another question that came to mind:

* Do we want the client to have retry on error behaviour? a maximum wait time per request?

Reply:
Open to suggestions, I also suggest we hardcode a sensible policy in. You can also come back to it later, a follow-up PR.

Reply:
Let's do that in another PR